tests: Remove stale symlinks to loop device on storage setup

aabawer opened the pull request:
os.symlink(device, link_path) may fail - so we need to delete stale symlink in setup.

This will happen since loop devices create for tests are not persistent between reboots,
so doing:

    python tests/storage.py setup
    reboot
    python tests/storage.py setup

Will result in stale symlinks for non-existing loop devices on last call
to storage setup.

nirs commented:
> Since symlinks are created under /var/tmp/ which is eventually cleared
> by the system, 

The system never clears /var/tmp.

> we may get to a point where the setup will fail for
> an already existing loop device but missing symlinks 

If we have a loop device without a symlink, it is not our loop device, so there is no problem.
Calling teardown will not do anything useful.

> and vice-versa.

If we have symlink without devices, this check in setup will fail:

 46     if not os.path.exists(LOOP1):                                                                                                                                        
 47         create_loop_device(LOOP1, BACKING1)

And setup will create a new loop device (good).

os.symlink(device, link_path) may fail - so we may need to delete stale symlink in setup.

I think this always happen now if you do:

1. python tests/storage.py setup
2. reboot
3. python tests/storage.py setup

> Adding a teardown before setup will assure we can take all required
> steps to make tests 4k ready.

This fix is not correct. The result is re-creating perfectly good setup for no reason, and having to enter the password again.

The correct fix is to delete stale symlink.

aabawer commented:
fixed the PR. thanks

nirs commented on tests/storage.py:
```diff
@@ -138,5 +140,15 @@ def remove_file(path):
+            os.unlink(path)
+    except EnvironmentError as e:
+        if e.errno != errno.ENOENT:
+            raise
```
I think we can simplify this a lot by adding this wrapper for creating symlink:

```
def create_symlink(src, dst):
    if os.path.islink(dst):
        os.unlink(dst)
    os.symlink(src, dst)
```

aabawer commented:
we have a precondition for all the loop device linking and making at line
48:
if not os.path.exists(LOOP1):
where LOOP1 is the link path, so its better to clean the dangling links
beforehand.

On Tue, May 28, 2019 at 4:51 PM Nir Soffer <notifications@github.com> wrote:

> *@nirs* requested changes on this pull request.
> ------------------------------
>
> In tests/storage.py
> <https://github.com/nirs/sanlock/pull/17#discussion_r288112237>:
>
> > @@ -138,5 +140,15 @@ def remove_file(path):
>              raise
>
>
> +def remove_stale_symlink(path):
> +    try:
> +        if not os.path.exists(os.readlink(path)):
> +            log.info("Removing stale symlink %s", path)
> +            os.unlink(path)
> +    except EnvironmentError as e:
> +        if e.errno != errno.ENOENT:
> +            raise
>
> I think we can simplify this a lot by adding this wrapper for creating
> symlink:
>
> def create_symlink(src, dst):
>     if os.path.islink(dst):
>         os.unlink(dst)
>     os.symlink(src, dst)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nirs/sanlock/pull/17?email_source=notifications&email_token=AL3BPNOCJI6YGPOOP2WFQPTPXU2HHA5CNFSM4HPWW2D2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOBZ3P4WI#pullrequestreview-242679385>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AL3BPNI2SBHOFHSNHF4MNEDPXU2HHANCNFSM4HPWW2DQ>
> .
>

nirs commented:
> we have a precondition for all the loop device linking and making at line 48: if not os.path.exists(LOOP1): where LOOP1 is the link path, so its better to clean the dangling links beforehand.

os.path.exist() check if the link target exist. When it return False we know that there is no loop device. We have two options:
- there is a stale link
- there is no link

os.path.islink() detect if we have a link that should be deleted.

Here is example flow:
```
>>> import os
>>> os.symlink("/no/such/path", "link")
>>> os.path.exists("link")
False
>>> os.path.islink("link")
True
>>> os.unlink("link")
>>> os.path.exists("link")
False
>>> os.path.islink("link")
False
```

aabawer committed:
tests: Remove stale symlinks before linking loop device

On storage setup we can have stale symlinks to loop devices
which were created before reboot and now are gone. Re-running the
setup will fail on creating the symlink again. In this patch we
add removal of the pre-existing symlink, if there is such, before
symlinking the new loop device.

aabawer commented:
Nice. Fixed the PR

On Tue, May 28, 2019 at 5:32 PM Nir Soffer <notifications@github.com> wrote:

> we have a precondition for all the loop device linking and making at line
> 48: if not os.path.exists(LOOP1): where LOOP1 is the link path, so its
> better to clean the dangling links beforehand.
>
> os.path.exist() check if the link target exist. When it return False we
> know that there is no loop device. We have two options:
>
>    - there is a stale link
>    - there is no link
>
> os.path.islink() detect if we have a link that should be deleted.
>
> Here is example flow:
>
> >>> import os
> >>> os.symlink("/no/such/path", "link")
> >>> os.path.exists("link")
> False
> >>> os.path.islink("link")
> True
> >>> os.unlink("link")
> >>> os.path.exists("link")
> False
> >>> os.path.islink("link")
> False
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nirs/sanlock/pull/17?email_source=notifications&email_token=AL3BPNISOLYV57LAOIZFZBLPXU65LA5CNFSM4HPWW2D2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWMKEYI#issuecomment-496542305>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AL3BPNMEUM7GQN6S2BHQJ6DPXU65LANCNFSM4HPWW2DQ>
> .
>

nirs reviewed (approved):
Thanks!
